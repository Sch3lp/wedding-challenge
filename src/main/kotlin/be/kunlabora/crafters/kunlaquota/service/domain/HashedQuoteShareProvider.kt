package be.kunlabora.crafters.kunlaquota.service.domain

class HashedQuoteShareProvider: QuoteShareProvider {
    private val alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"

    override fun invoke(quoteId: QuoteId): QuoteShare =
        quoteId.value.take(8)
            .hexToNumber().toLong()
            .encodeToBase26()
            .asQuoteShare()

    private fun String.asQuoteShare() = QuoteShare(this)

    @OptIn(ExperimentalStdlibApi::class)
    private fun String.hexToNumber(): String = hexToUInt(HexFormat.Default).toString()

    private fun Long.encodeToBase26(): String {
        val base = alphabet.length
        var num = this
        val result = StringBuilder()

        return if (num == 0L) {
            alphabet[0].toString()
        } else {
            while (num > 0) {
                val remainder = (num % base).toInt()
                result.append(alphabet[remainder])
                num /= base
            }

            result.reverse().toString()
        }
    }
}