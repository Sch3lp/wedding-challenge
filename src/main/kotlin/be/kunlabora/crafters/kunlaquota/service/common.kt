package be.kunlabora.crafters.kunlaquota.service

import be.kunlabora.crafters.kunlaquota.AddFailure
import be.kunlabora.crafters.kunlaquota.QuoteIsInvalid
import be.kunlabora.crafters.kunlaquota.service.Result.Error
import be.kunlabora.crafters.kunlaquota.service.Result.Ok
import be.kunlabora.crafters.kunlaquota.service.domain.Quote
import be.kunlabora.crafters.kunlaquota.service.domain.QuoteId
import com.fasterxml.jackson.annotation.JsonTypeInfo
import java.util.*

typealias IdProvider = () -> String

val uuidStringProvider: IdProvider = { UUID.randomUUID().toString() }

@Suppress("unused")
class EntityId<E> private constructor(val value: String) {
    companion object {
        fun <E> new(idProvider: IdProvider = uuidStringProvider): EntityId<E> =
            EntityId(idProvider())

        fun <E> fromString(value: String): EntityId<E> = EntityId(UUID.fromString(value).toString())
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as EntityId<*>

        return value == other.value
    }

    override fun hashCode(): Int = value.hashCode()

    override fun toString() = "EntityId($value)"
}

//marker interface
@JsonTypeInfo(use = JsonTypeInfo.Id.NAME)
sealed interface Command
data class AddQuote(val lines: List<Quote.Line>) : Command {

    fun validate(): Result<AddFailure, AddQuote> = validateLines().map { clean() }

    private fun validateLines() = when {
            onlyHasEmptyLines() -> Error(QuoteIsInvalid("A Quote needs at least one Line."))
            someLinesHaveSameOrder() -> Error(QuoteIsInvalid("Can't have multiple lines with the same order."))
            someLinesHaveNoName() -> Error(QuoteIsInvalid("A Quote Line needs a name."))
            someLinesHaveNoText() -> Error(QuoteIsInvalid("A Quote Line needs text."))
            else -> Ok(this)
        }

    private fun AddQuote.onlyHasEmptyLines() =
        lines.filterNot { it.text.isBlank() && it.name.isBlank() }.isEmpty()

    private fun AddQuote.someLinesHaveSameOrder() =
        lines.map { it.order }.toSet().size < lines.size

    private fun AddQuote.someLinesHaveNoName() =
        lines.any { line -> line.name.isBlank() && line.text.isNotBlank() }

    private fun AddQuote.someLinesHaveNoText() =
        lines.any { line -> line.text.isBlank() && line.name.isNotBlank() }

    private fun clean(): AddQuote = copy(
        lines = lines.dropBlankLines().reindex()
    )

    private fun List<Quote.Line>.dropBlankLines() =
        filterNot { it.name.isBlank() && it.text.isBlank() }

    private fun List<Quote.Line>.reindex() =
        sortedBy { it.order }
            .mapIndexed { index, line -> line.copy(order = index + 1) }
}

data class ShareQuote(val id: QuoteId) : Command

sealed class Result<out F, out T> {
    data class Error<F>(val value: F) : Result<F, Nothing>()
    data class Ok<T>(val value: T) : Result<Nothing, T>()

    fun valueOrThrow() = when (this) {
        is Error -> error("You expected an Ok, but got an Error of ${this.value}")
        is Ok -> this.value
    }

    fun <R> map(block: (T) -> R): Result<F, R> =
        when (this) {
            is Error -> this
            is Ok -> Ok(block(this.value))
        }

    fun <R> flatMap(block: (T) -> Result<@UnsafeVariance F, R>): Result<F, R> =
        when (this) {
            is Error -> this
            is Ok -> block(this.value)
        }

    fun recover(recovery: (F) -> @UnsafeVariance T): Result<T, T> =
        when (this) {
            is Error -> Ok(recovery(this.value))
            is Ok -> this
        }
}


fun <T> Result<T, T>.get(): T =
    when (this) {
        is Error -> this.value
        is Ok -> this.value
    }
