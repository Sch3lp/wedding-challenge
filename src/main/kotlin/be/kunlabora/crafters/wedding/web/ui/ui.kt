package be.kunlabora.crafters.wedding.web.ui

import be.kunlabora.crafters.wedding.service.domain.Assignee
import be.kunlabora.crafters.wedding.web.ui.components.Hero.hero
import kotlinx.html.*
import kotlinx.html.stream.appendHTML
import java.io.StringWriter

fun wrapper(title: String, selectedAssignee: Assignee?, block: BODY.() -> Unit) =
    StringWriter().appendHTML().html {
        head {
            title { +title }
            meta(charset = "utf-8")
            meta(name = "viewport", content = "width=device-width, initial-scale=1")
            link(
                rel = "stylesheet",
                href = "https://cdn.jsdelivr.net/npm/bulma@1.0.1/css/versions/bulma-no-dark-mode.min.css"
            )
            link(
                rel = "stylesheet",
                href = "https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free@6.4.2/css/fontawesome.min.css"
            )
            script(src = "https://unpkg.com/htmx.org@2.0.0") {}
            script(src = "https://unpkg.com/hyperscript.org@0.9.12") {}
        }
        body {
            hero(selectedAssignee)
            div(classes = "block") { id = "errorMessages" }
            block()
            div { id = "modals-here" }
        }
    }.toString()

fun partial(block: FlowContent.() -> Unit): String {
    val writer = StringWriter()
    val consumer = writer.appendHTML()
    // hacky stuff so we don't have to return a wrapper div
    object : FlowContent {
        override val consumer = consumer
        override val attributes: MutableMap<String, String>
            get() = mutableMapOf()
        override val attributesEntries: Collection<Map.Entry<String, String>>
            get() = emptyList()
        override val emptyTag: Boolean
            get() = true
        override val inlineTag: Boolean
            get() = true
        override val namespace: String?
            get() = null
        override val tagName: String
            get() = ""
    }.block()
    return writer.toString()
}

fun String.path(vararg paths: String): String =
    buildList {
        add(this@path)
        addAll(paths)
    }.joinToString("/")