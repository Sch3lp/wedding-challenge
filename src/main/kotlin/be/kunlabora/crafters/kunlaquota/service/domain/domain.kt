package be.kunlabora.crafters.kunlaquota.service.domain

import be.kunlabora.crafters.kunlaquota.service.EntityId
import java.time.LocalDateTime

typealias QuoteId = EntityId<Quote>

data class Quote(
    val id: QuoteId,
    val lines: List<Line> = emptyList(),
    val at: LocalDateTime = LocalDateTime.now(),
) {
    data class Line(
        val order: Int,
        val name: String,
        val text: String,
    )
}

@JvmInline value class QuoteShare(val value: String)