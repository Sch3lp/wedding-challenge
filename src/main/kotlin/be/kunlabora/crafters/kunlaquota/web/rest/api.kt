package be.kunlabora.crafters.kunlaquota.web.rest

import be.kunlabora.crafters.kunlaquota.AddFailure
import be.kunlabora.crafters.kunlaquota.QuoteAlreadyExists
import be.kunlabora.crafters.kunlaquota.QuoteIsInvalid
import be.kunlabora.crafters.kunlaquota.ShareQuoteFailed
import be.kunlabora.crafters.kunlaquota.service.*
import be.kunlabora.crafters.kunlaquota.service.domain.QuoteShare
import org.slf4j.LoggerFactory
import org.springframework.http.MediaType
import org.springframework.web.servlet.function.*
import java.net.URI

fun apiRoutes(quotes: IQuotes): RouterFunctionDsl.() -> Unit = {
    fun AddQuote.execute() = quotes.execute(this)
    fun ShareQuote.execute() = quotes.execute(this)

    "/quote".nest {
        GET(RequestPredicates.param("share") { true }) { request ->
            quotes.findByQuoteShare(QuoteShare(request.paramOrNull("share")!!))
                .map { foundQuotes ->
                    ServerResponse.ok()
                        .contentType(MediaType.APPLICATION_JSON)
                        .body(foundQuotes)
                }
                .recover { failure ->
                    ServerResponse.status(404).build().also { logger.error("$failure") }
                }.get()
        }

        GET("") {
            quotes.findAll()
                .map { foundQuotes ->
                    ServerResponse.ok()
                        .contentType(MediaType.APPLICATION_JSON)
                        .body(foundQuotes)
                }
                .recover { failure ->
                    ServerResponse.status(500).build().also { logger.error("$failure") }
                }.get()
        }

        POST("") { request ->
            request.body<AddQuote>()
                .execute()
                .map { quote ->
                    ServerResponse.created(quote.id.asUri(request)).build()
                }
                .handleAddQuoteFailure()
                .get()
        }

        POST("share") { request ->
            request.body<ShareQuote>()
                .execute()
                .map { quoteShare ->
                    ServerResponse.ok().body(quoteShare.asUri(request).toString())
                }
                .recover { failure ->
                    when (failure) {
                        is ShareQuoteFailed -> ServerResponse.status(500).build().also { logger.error(failure.message) }
                    }
                }
                .get()
        }
    }
}

private fun Result<AddFailure, ServerResponse>.handleAddQuoteFailure() =
    recover { failure ->
        when (failure) {
            is QuoteIsInvalid -> ServerResponse.status(400).body(AddQuoteError(failure.message))
            is QuoteAlreadyExists -> ServerResponse.status(500)
                .body(AddQuoteError("Something fishy occurred on our side. Maybe try again?"))
        }.also { logger.error(failure.message) }
    }


private val logger = LoggerFactory.getLogger("KunlaQuotaLogger")

private fun <E> EntityId<E>.asUri(request: ServerRequest): URI =
    request.uriBuilder().path("/{id}").build(this.value)

private fun QuoteShare.asUri(request: ServerRequest): URI =
    request.uriBuilder().replacePath("/api/quote").queryParam("share",this.value).build()