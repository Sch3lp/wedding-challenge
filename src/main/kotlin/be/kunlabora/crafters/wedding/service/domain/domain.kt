package be.kunlabora.crafters.wedding.service.domain

import be.kunlabora.crafters.wedding.service.EntityId
import java.time.LocalDateTime

typealias ChallengeId = EntityId<Challenge>
typealias AssigneeId = EntityId<Assignee>

data class Challenge(
    val id: ChallengeId = ChallengeId.new(),
    val description: String,
    val assignee: AssigneeId,
    val completed: Boolean = false,
    val completedOn: LocalDateTime? = null,
) {
    fun markAsCompleted(): Challenge =
        copy(completed = true, completedOn = LocalDateTime.now())
}

data class Assignee(
    val name: String,
    val id: AssigneeId = AssigneeId.new(),
)