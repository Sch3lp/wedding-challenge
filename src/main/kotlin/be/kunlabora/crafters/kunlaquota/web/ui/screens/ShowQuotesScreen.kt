package be.kunlabora.crafters.kunlaquota.web.ui.screens

import be.kunlabora.crafters.kunlaquota.service.domain.Quote
import be.kunlabora.crafters.kunlaquota.web.ui.baseUiUrl
import be.kunlabora.crafters.kunlaquota.web.ui.components.Htmx.hxGet
import be.kunlabora.crafters.kunlaquota.web.ui.components.Htmx.hxTarget
import be.kunlabora.crafters.kunlaquota.web.ui.components.Htmx.hyper
import be.kunlabora.crafters.kunlaquota.web.ui.components.Util.formatToKunlaDate
import be.kunlabora.crafters.kunlaquota.web.ui.path
import be.kunlabora.crafters.kunlaquota.web.ui.screens.AddQuoteModal.addQuoteModalId
import kotlinx.html.*
import java.time.LocalDateTime

object ShowQuotesScreen {
    fun FlowContent.showQuotes(allQuotes: List<Quote>) {
        section(classes = "section") {
            id = "quote-section"
            div(classes = "container") {
                if (allQuotes.isEmpty()) {
                    + "Be the first to create a new quote!"
                }else allQuotes.forEach { quote(it) }
            }
        }

        button(classes = "button is-primary") {
            id = "add-quote"
            style = "position: fixed; bottom: 20px; right: 20px;"
            hxGet = baseUiUrl.path("showModal")
            hxTarget = "#modals-here"
            hyper = "on htmx:afterOnLoad wait 10ms then add .is-active to $addQuoteModalId"
            +"Add Quote"
        }
    }

    private fun FlowContent.quote(quote: Quote) {
        div("card") {
            div("card-content") {
                quote.lines.forEach { quoteLine(it); br }
                p(classes = "subtitle") {
                    quoteDate(quote.at)
                }
            }
        }
    }

    private fun FlowContent.quoteLine(line: Quote.Line) {
        p {
            strong { +"${line.name}: " }
            +line.text
        }
    }

    private fun FlowContent.quoteDate(at: LocalDateTime) {
        div("is-flex is-justify-content-flex-end") {
            p("is-size-7 has-text-grey-light") {
                +at.formatToKunlaDate()
            }
        }
    }

    fun FlowContent.errorMessage(vararg messages: String) {
        article("message is-danger") {
            div(classes = "message-body") {
                messages.forEach { message ->
                    p { +message }
                }
            }
        }
    }
}