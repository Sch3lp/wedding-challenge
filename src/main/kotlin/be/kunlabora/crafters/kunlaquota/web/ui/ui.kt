package be.kunlabora.crafters.kunlaquota.web.ui

import be.kunlabora.crafters.kunlaquota.service.AddQuote
import be.kunlabora.crafters.kunlaquota.service.IQuotes
import be.kunlabora.crafters.kunlaquota.service.domain.Quote
import be.kunlabora.crafters.kunlaquota.service.domain.QuoteShare
import be.kunlabora.crafters.kunlaquota.service.get
import be.kunlabora.crafters.kunlaquota.web.ui.components.Hero.hero
import be.kunlabora.crafters.kunlaquota.web.ui.screens.AddQuoteModal.addQuoteLine
import be.kunlabora.crafters.kunlaquota.web.ui.screens.AddQuoteModal.addQuoteModal
import be.kunlabora.crafters.kunlaquota.web.ui.screens.ShowQuotesScreen.errorMessage
import be.kunlabora.crafters.kunlaquota.web.ui.screens.ShowQuotesScreen.showQuotes
import kotlinx.html.*
import kotlinx.html.stream.appendHTML
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType
import org.springframework.web.servlet.function.RequestPredicates
import org.springframework.web.servlet.function.RouterFunctionDsl
import org.springframework.web.servlet.function.ServerResponse
import org.springframework.web.servlet.function.paramOrNull
import java.io.StringWriter

const val baseUiUrl = "/ui"

fun uiRoutes(quotes: IQuotes): RouterFunctionDsl.() -> Unit = {
    GET("") {
        val title = "KunlaQuota"
        ServerResponse.status(HttpStatus.OK)
            .contentType(MediaType.TEXT_HTML)
            .body(
                wrapper(title) {
                    quotes.findAll()
                        .map { quotes -> showQuotes(quotes) }
                }
            )
    }
    GET("showModal") {
        ServerResponse.status(HttpStatus.OK)
            .contentType(MediaType.TEXT_HTML)
            .body(
                partial {
                    addQuoteModal()
                }
            )
    }
    POST("new/addLine") {
        ServerResponse.status(HttpStatus.OK)
            .body(
                partial { addQuoteLine() }
            )
    }
    POST("new") { request ->
        val names = request.params()["nameInput"] ?: emptyList()
        val texts = request.params()["textInput"] ?: emptyList()

        val addQuote = AddQuote(
            names.zip(texts).mapIndexed { idx, (name, text) -> Quote.Line(idx, name, text) }
        )
        quotes.execute(addQuote)
            .map {
                ServerResponse.status(HttpStatus.OK)
                    .header("HX-Redirect", baseUiUrl)
                    .build()
            }.recover { failure ->
                ServerResponse.status(HttpStatus.OK)
                    .body(partial { errorMessage("Oopsie! Something broke!", failure.message) })
            }.get()
    }
    GET(RequestPredicates.param("share") { true }) { request ->
        val quoteShare = QuoteShare(request.paramOrNull("share")!!)
        quotes.findByQuoteShare(quoteShare)
            .map { quotes ->
                ServerResponse.status(HttpStatus.OK)
                    .body(wrapper("Shared quote: ${quoteShare.value}") {
                        showQuotes(quotes)
                    })
            }.recover { failure ->
                ServerResponse.status(HttpStatus.OK)
                    .body(partial { errorMessage("Oopsie! Something broke!", failure.message) })
            }.get()
    }
}

fun wrapper(title: String, block: BODY.() -> Unit) =
    StringWriter().appendHTML().html {
        head {
            title { +title }
            meta(charset = "utf-8")
            meta(name = "viewport", content = "width=device-width, initial-scale=1")
            link(
                rel = "stylesheet",
                href = "https://cdn.jsdelivr.net/npm/bulma@1.0.1/css/versions/bulma-no-dark-mode.min.css"
            )
            link(
                rel = "stylesheet",
                href = "https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free@6.4.2/css/fontawesome.min.css"
            )
            script(src = "https://unpkg.com/htmx.org@2.0.0") {}
            script(src = "https://unpkg.com/hyperscript.org@0.9.12") {}
        }
        body {
            hero()
            div(classes = "block") { id = "errorMessages" }
            block()
            div { id = "modals-here" }
        }
    }.toString()

fun partial(block: FlowContent.() -> Unit): String {
    val writer = StringWriter()
    val consumer = writer.appendHTML()
    // hacky stuff so we don't have to return a wrapper div
    object : FlowContent {
        override val consumer = consumer
        override val attributes: MutableMap<String, String>
            get() = mutableMapOf()
        override val attributesEntries: Collection<Map.Entry<String, String>>
            get() = emptyList()
        override val emptyTag: Boolean
            get() = true
        override val inlineTag: Boolean
            get() = true
        override val namespace: String?
            get() = null
        override val tagName: String
            get() = ""
    }.block()
    return writer.toString()
}

fun String.path(vararg paths: String): String =
    buildList {
        add(this@path)
        addAll(paths)
    }.joinToString("/")