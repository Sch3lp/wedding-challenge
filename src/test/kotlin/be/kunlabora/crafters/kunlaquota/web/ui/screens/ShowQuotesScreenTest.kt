package be.kunlabora.crafters.kunlaquota.web.ui.screens

import be.kunlabora.crafters.kunlaquota.service.aMultiLineQuote
import be.kunlabora.crafters.kunlaquota.service.aSingleLineQuote
import be.kunlabora.crafters.kunlaquota.service.domain.Quote
import be.kunlabora.crafters.kunlaquota.web.ui.partial
import be.kunlabora.crafters.kunlaquota.web.ui.screens.ShowQuotesScreen.showQuotes
import org.junit.jupiter.api.Test
import verify

class ShowQuotesScreenTest {

    @Test
    fun `showQuotes when no quotes found`() {
        val quotes = emptyList<Quote>()
        partial { showQuotes(quotes) }.verify()
    }

    @Test
    fun `showQuotes when some quotes found`() {
        val quotes = buildList<Quote> {
            aMultiLineQuote {
                "Charlie" said "Liam, my brother, just had a daughter. So you know, we're taking some time off."
                "Charlie" said "Liam and I don't keep in touch and there's been some trouble with... royalties."
                "Charlie" said "The fact of the matter sir is that. DriveShaft is... dead."
            }
            aSingleLineQuote(name = "Jooones", text = "snorf snorf?")
        }
        partial { showQuotes(quotes) }.verify()
    }
}