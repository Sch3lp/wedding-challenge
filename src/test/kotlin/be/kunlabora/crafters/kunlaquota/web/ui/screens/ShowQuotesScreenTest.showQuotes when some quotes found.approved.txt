<section class="section" id="quote-section">
  <div class="container">Be the first to create a new quote!</div>
</section>
<button class="button is-primary" id="add-quote" style="position: fixed; bottom: 20px; right: 20px;" hx-get="/ui/showModal" hx-target="#modals-here" _="on htmx:afterOnLoad wait 10ms then add .is-active to addquotemodal">Add Quote</button>